packages/orm/src/Model.ts

```diff
@@ -0,0 +1,786 @@
+        // If a value is supplied for a ManyToMany field,
+        // discard them from props and save for later processing.
+        m2mRelations[key] = userProps[key];
+        delete props[key];
+      }
+    });
+
+    // add backward many-many if required
+    declaredVirtualFieldNames.forEach((key) => {
+      if (!m2mRelations.hasOwnProperty(key)) {
+        const field = this.virtualFields[key];
+        if (userProps.hasOwnProperty(key) && field instanceof ManyToMany) {
+          // If a value is supplied for a ManyToMany field,
+          // discard them from props and save for later processing.
+          m2mRelations[key] = userProps[key];
+          delete props[key];
+        }
+      }
+    });
+
+    const newEntry = this.session.applyUpdate({
+      action: CREATE,
+      table: this.modelName,
+      payload: props,
+    });
+
+    const ThisModel = this;
+    const instance = new ThisModel(newEntry);
+    instance._refreshMany2Many(m2mRelations); // eslint-disable-line no-underscore-dangle
+    return instance;
+  }
+
+  /**
+   * Creates a new or update existing record in the database, instantiates a {@link Model} and returns it.
+   *
+   * If you pass values for many-to-many fields, instances are created on the through
+   * model as well.
+   *
+   * @param  {props} userProps - the required {@link Model}'s properties.
+   * @return {Model} a {@link Model} instance.
+   */
+  static upsert(userProps: { [k: string]: any }) {
+    if (typeof this.session === "undefined") {
+      throw new Error(
+        [
+          `Tried to upsert a ${this.modelName} model instance without a session. `,
+          "Create a session using `session = orm.session()` and call ",
+          `\`session["${this.modelName}"].upsert\` instead.`,
+        ].join("")
+      );
+    }
+
+    const { idAttribute } = this;
+    if (userProps.hasOwnProperty(idAttribute)) {
+      const id = userProps[idAttribute];
+      if (this.idExists(id)) {
+        const model = this.withId(id) as Model;
+        model.update(userProps);
+        return model;
+      }
+    }
+
+    return this.create(userProps);
+  }
+
+  /**
+   * Returns a {@link Model} instance for the object with id `id`.
+   * Returns `null` if the model has no instance with id `id`.
+   *
+   * You can use {@link Model#idExists} to check for existence instead.
+   *
+   * @param  {*} id - the `id` of the object to get
+   * @throws If object with id `id` doesn't exist
+   * @return {Model|null} {@link Model} instance with id `id`
+   */
+  static withId(id: ModelId) {
+    return this.get({
+      [this.idAttribute]: id,
+    });
+  }
+
+  /**
+   * Returns a boolean indicating if an entity
+   * with the id `id` exists in the state.
+   *
+   * @param  {*}  id - a value corresponding to the id attribute of the {@link Model} class.
+   * @return {Boolean} a boolean indicating if entity with `id` exists in the state
+   *
+   * @since 0.11.0
+   */
+  static idExists(id?: any) {
```

Comment:
In general I think we should replace all `any` with `unknown` - if we can't specify the types